workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/HandlePlayerHeartbeatResponseSystem.cs

```diff
@@ -4,87 +4,78 @@
-                        {
-                            PostUpdateCommands.SetComponent(entity, failedHeartbeatsComponent);
-                        }
-                        else
-                        {
-                            PostUpdateCommands.AddComponent(entity, failedHeartbeatsComponent);
-                        }
+                        stillAlive = true;
+                        break;
+                    }
+                }
 
-                        Debug.LogFormat(Messages.FailedHeartbeat, entityId,
-                            failedHeartbeatsComponent.NumFailedHeartbeats,
-                            PlayerLifecycleConfig.MaxNumFailedPlayerHeartbeats);
+                if (stillAlive)
+                {
+                    heartbeatsComponent.NumFailedHeartbeats = 0;
+                }
+                else
+                {
+                    heartbeatsComponent.NumFailedHeartbeats++;
 
-                        if (failedHeartbeatsComponent.NumFailedHeartbeats >=
-                            PlayerLifecycleConfig.MaxNumFailedPlayerHeartbeats)
-                        {
-                            Debug.LogFormat(Messages.DeletingPlayer, entityId);
-                            entityDeleteSender.RequestsToSend.Add(WorldCommands.DeleteEntity.CreateRequest
-                            (
-                                entityId
-                            ));
-                            data.WorldCommandSenders[i] = entityDeleteSender;
-                        }
-                    }
-                    else
+                    if (heartbeatsComponent.NumFailedHeartbeats >=
+                        PlayerLifecycleConfig.MaxNumFailedPlayerHeartbeats)
                     {
-                        if (EntityManager.HasComponent<HeartbeatData>(entity))
-                        {
-                            PostUpdateCommands.RemoveComponent<HeartbeatData>(entity);
-                        }
+                        var entityDeleteSender = heartbeatGroup.WorldCommandSenders[i];
+                        entityDeleteSender.RequestsToSend.Add(WorldCommands.DeleteEntity.CreateRequest
+                        (
+                            heartbeatGroup.SpatialEntityIds[i].EntityId
+                        ));
+                        heartbeatGroup.WorldCommandSenders[i] = entityDeleteSender;
                     }
                 }
 
-                PostUpdateCommands.RemoveComponent<AwaitingHeartbeatResponseTag>(entity);
+                heartbeatGroup.Heartbeats[i] = heartbeatsComponent;
             }
         }
-
-        internal static class Messages
-        {
-            public const string FailedHeartbeat =
-                "Client of player entity {0} failed to respond to heartbeat. Number of consecutive failures: {1}/{2}.";
-
-            public const string DeletingPlayer =
-                "Client of player entity {0} failed to respond to heartbeat and exhausted all retries. Deleting player.";
```

Comment:
+1. This feature module should log what it is doing. Otherwise it will confuse new users. Eventually we can have an option to set a log level per feature module.